### PR TITLE
feat(oft-solana): introduce addComputeUnitInstructions

### PR DESCRIPTION
**Pull request by ryandgoulding**

### CODE REVIEW

**Comment by ryandgoulding** [comment]
not ready for review, just letting it go through CI pipeline as it is faster than doing so locally

**Comment by DanL0** [comment]
@ryandgoulding can you resolve conflict?

**Comment by ryandgoulding** [comment]
@DanL0 done, can you take a look?

File: examples/oft-solana/tasks/solana/index.ts
Comment (DanL0) [inline examples/oft-solana/tasks/solana/index.ts:164]: Error:

image

Seems running locally:
```
pnpm tsc
```

should catch this type of error

```
➜  oapp-prio-fees git:(master) ✗ pnpm hardhat lz:oft:solana:create --eid 40168 --program-id 3ZuKaovZkFLf5StLP2sHDBa65WKR7fjD1AViFugCFqRH --amount 77777777123 --only-oft-store true
bigint: Failed to load bindings, pure JS will be used (try npm run rebuild?)
No additional minters specified.  This will result in only the OFTStore being able to mint new tokens.
created SPL multisig @ APA9owkDi6SUUgXFbG5SaHb5nT8TpadA3mWK23DY6AzQ
Error fetching prioritization fees: TypeError: Cannot read properties of undefined (reading '_bn')
    at isPublicKeyData (/Users/danielkmak/projects/oapps/oapp-prio-fees/node_modules/.pnpm/@solana+web3.js@1.95.8/node_modules/@solana/web3.js/src/publickey.ts:39:35)
    at new PublicKey (/Users/danielkmak/projects/oapps/oapp-prio-fees/node_modules/.pnpm/@solana+web3.js@1.95.8/node_modules/@solana/web3.js/src/publickey.ts:58:9)
    at getPrioritizationFees (/Users/danielkmak/projects/oapps/oapp-prio-fees/tasks/utils/getFee.ts:22:27)
    at getComputeUnitPriceAndLimit (/Users/danielkmak/projects/oapps/oapp-prio-fees/tasks/solana/index.ts:164:54)
    at addPerformanceInstructions (/Users/danielkmak/projects/oapps/oapp-prio-fees/tasks/solana/index.ts:195:81)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async SimpleTaskDefinition.action (/Users/danielkmak/projects/oapps/oapp-prio-fees/tasks/solana/createOFT.ts:231:29)
    at async Environment._runTaskDefinition (/Users/danielkmak/projects/oapps/oapp-prio-fees/node_modules/.pnpm/hardhat@2.22.17_ts-node@10.9.2_typescript@5.7.2/node_modules/hardhat/src/internal/core/runtime-environment.ts:351:14)
    at async Environment.run (/Users/danielkmak/projects/oapps/oapp-prio-fees/node_modules/.pnpm/hardhat@2.22.17_ts-node@10.9.2_typescript@5.7.2/node_modules/hardhat/src/internal/core/runtime-environment.ts:184:14)
    at async main (/Users/danielkmak/projects/oapps/oapp-prio-fees/node_modules/.pnpm/hardhat@2.22.17_ts-node@10.9.2_typescript@5.7.2/node_modules/hardhat/src/internal/cli/cli.ts:324:7)
createTokenTx: https://explorer.solana.com/tx/4BuTsrPvAx1687uvFvvjZLHs9Erorwi5WnNwFbE4TMwAV3sDhRs4CfmH4Q6bkg4QDYTTKV9xqN2pERjsmFbXhBDG?cluster=devnet
Error fetching prioritization fees: TypeError: Cannot read properties of undefined (reading '_bn')
    at isPublicKeyData (/Users/danielkmak/projects/oapps/oapp-prio-fees/node_modules/.pnpm/@solana+web3.js@1.95.8/node_modules/@solana/web3.js/src/publickey.ts:39:35)
    at new PublicKey (/Users/danielkmak/projects/oapps/oapp-prio-fees/node_modules/.pnpm/@solana+web3.js@1.95.8/node_modules/@solana/web3.js/src/publickey.ts:58:9)
    at getPrioritizationFees (/Users/danielkmak/projects/oapps/oapp-prio-fees/tasks/utils/getFee.ts:22:27)
    at getComputeUnitPriceAndLimit (/Users/danielkmak/projects/oapps/oapp-prio-fees/tasks/solana/index.ts:164:54)
    at addPerformanceInstructions (/Users/danielkmak/projects/oapps/oapp-prio-fees/tasks/solana/index.ts:195:81)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async SimpleTaskDefinition.action (/Users/danielkmak/projects/oapps/oapp-prio-fees/tasks/solana/createOFT.ts:262:25)
    at async Environment._runTaskDefinition (/Users/danielkmak/projects/oapps/oapp-prio-fees/node_modules/.pnpm/hardhat@2.22.17_ts-node@10.9.2_typescript@5.7.2/node_modules/hardhat/src/internal/core/runtime-environment.ts:351:14)
    at async Environment.run (/Users/danielkmak/projects/oapps/oapp-prio-fees/node_modules/.pnpm/hardhat@2.22.17_ts-node@10.9.2_typescript@5.7.2/node_modules/hardhat/src/internal/core/runtime-environment.ts:184:14)
    at async main (/Users/danielkmak/projects/oapps/oapp-prio-fees/node_modules/.pnpm/hardhat@2.22.17_ts-node@10.9.2_typescript@5.7.2/node_modules/hardhat/src/internal/cli/cli.ts:324:7)
initOftTx: https://explorer.solana.com/tx/3Xs97mZUvXpq2V1VKiBAF7aaATgEXgDhxGXWbv21DXHw2J6uyre1z9SBZhmteHTf56TnJ8RuDR4V2GSpogSZsd4T?cluster=devnet
Error fetching prioritization fees: TypeError: Cannot read properties of undefined (reading '_bn')
    at isPublicKeyData (/Users/danielkmak/projects/oapps/oapp-prio-fees/node_modules/.pnpm/@solana+web3.js@1.95.8/node_modules/@solana/web3.js/src/publickey.ts:39:35)
    at new PublicKey (/Users/danielkmak/projects/oapps/oapp-prio-fees/node_modules/.pnpm/@solana+web3.js@1.95.8/node_modules/@solana/web3.js/src/publickey.ts:58:9)
    at getPrioritizationFees (/Users/danielkmak/projects/oapps/oapp-prio-fees/tasks/utils/getFee.ts:22:27)
    at getComputeUnitPriceAndLimit (/Users/danielkmak/projects/oapps/oapp-prio-fees/tasks/solana/index.ts:164:54)
    at addPerformanceInstructions (/Users/danielkmak/projects/oapps/oapp-prio-fees/tasks/solana/index.ts:195:81)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async SimpleTaskDefinition.action (/Users/danielkmak/projects/oapps/oapp-prio-fees/tasks/solana/createOFT.ts:292:29)
    at async Environment._runTaskDefinition (/Users/danielkmak/projects/oapps/oapp-prio-fees/node_modules/.pnpm/hardhat@2.22.17_ts-node@10.9.2_typescript@5.7.2/node_modules/hardhat/src/internal/core/runtime-environment.ts:351:14)
    at async Environment.run (/Users/danielkmak/projects/oapps/oapp-prio-fees/node_modules/.pnpm/hardhat@2.22.17_ts-node@10.9.2_typescript@5.7.2/node_modules/hardhat/src/internal/core/runtime-environment.ts:184:14)
    at async main (/Users/danielkmak/projects/oapps/oapp-prio-fees/node_modules/.pnpm/hardhat@2.22.17_ts-node@10.9.2_typescript@5.7.2/node_modules/hardhat/src/internal/cli/cli.ts:324:7)
setAuthorityTx: https://explorer.solana.com/tx/5QtuXHuANz4XAQiDJRwaJmjtjkTLHyyTStrh7dzZ6g7RFNutbHqF554BxEgATenpoqL8rZNu2NH3B5hwcyVdyPB?cluster=devnet
Accounts have been saved to ./deployments/solana-testnet/OFT.json
```

**Review by DanL0 — DISMISSED** [review]

File: examples/oft-solana/tasks/solana/index.ts
Comment (nazreen) [inline examples/oft-solana/tasks/solana/index.ts:185]: can we call them `addComputeUnitInstructions` instead? It's more descriptive that way.

Comment (St0rmBr3w) [reply examples/oft-solana/tasks/solana/index.ts:185]: I agree, keeping it inline with the actual lingo for Solana is better.

File: examples/oft-solana/tasks/solana/index.ts
Comment (ryandgoulding) [reply examples/oft-solana/tasks/solana/index.ts:164]: Got it, the internal API signature changed in #1084.  Will update it.

File: examples/oft-solana/tasks/solana/index.ts
Comment (ryandgoulding) [reply examples/oft-solana/tasks/solana/index.ts:185]: Sure.  I am not a great namer.

File: examples/oft-solana/tasks/solana/index.ts
Comment (nazreen) [inline examples/oft-solana/tasks/solana/index.ts:39]: question: what addresses are in these lookup tables ?

**Comment by nazreen** [comment]
I tested locally and here are my results, using the `time` utility in the terminal to measure

task ran: createOFT

without calling addPerformanceInstructions
```
pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  2.02s user 0.34s system 4% cpu 53.281 total
pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  1.99s user 0.35s system 4% cpu 52.599 total
pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  2.04s user 0.37s system 4% cpu 53.974 total
pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  1.83s user 0.32s system 5% cpu 40.226 total
pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true   1.95s user 0.34s system 4% cpu 52.211 total
pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true   2.02s user 0.34s system 4% cpu 53.015 total
```

with calling addPerformanceInstructions
```
pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  2.06s user 0.33s system 5% cpu 42.535 total
pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  2.04s user 0.37s system 4% cpu 53.974 total
pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  2.17s user 0.37s system 3% cpu 1:04.41 total
pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  2.27s user 0.36s system 4% cpu 1:05.64 total
pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  2.05s user 0.35s system 5% cpu 43.787 total
pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  1.99s user 0.34s system 5% cpu 39.079 total
```

So overall it does improve transactions processing times

@ryandgoulding we can merge it, just need to fix the `getFee` interface change and rename the function

If you want me to help, I don't mind

**Comment by ryandgoulding** [comment]
> I tested locally and here are my results, using the `time` utility in the terminal to measure
> 
> task ran: createOFT
> 
> without calling addPerformanceInstructions
> 
> ```
> pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  2.02s user 0.34s system 4% cpu 53.281 total
> pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  1.99s user 0.35s system 4% cpu 52.599 total
> pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  2.04s user 0.37s system 4% cpu 53.974 total
> pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  1.83s user 0.32s system 5% cpu 40.226 total
> pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true   1.95s user 0.34s system 4% cpu 52.211 total
> pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true   2.02s user 0.34s system 4% cpu 53.015 total
> ```
> 
> with calling addPerformanceInstructions
> 
> ```
> pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  2.06s user 0.33s system 5% cpu 42.535 total
> pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  2.04s user 0.37s system 4% cpu 53.974 total
> pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  2.17s user 0.37s system 3% cpu 1:04.41 total
> pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  2.27s user 0.36s system 4% cpu 1:05.64 total
> pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  2.05s user 0.35s system 5% cpu 43.787 total
> pnpm hardhat lz:oft:solana:create --eid 40168 --program-id $PROGRAM_ID  true  1.99s user 0.34s system 5% cpu 39.079 total
> ```
> 
> So overall it does improve transactions processing times
> 
> @ryandgoulding we can merge it, just need to fix the `getFee` interface change and rename the function
> 
> If you want me to help, I don't mind

Sure, please feel free to take over the branch!

File: examples/oft-solana/tasks/solana/index.ts
Comment (nazreen) [reply examples/oft-solana/tasks/solana/index.ts:39]: Thanks. I was wondering what accounts are these addresses representing ? Would be good to document somewhere.

File: examples/oft-solana/tasks/solana/index.ts
Comment (nazreen) [reply examples/oft-solana/tasks/solana/index.ts]: renamed in  https://github.com/LayerZero-Labs/devtools/pull/1091/commits/294318a3e59efd4374cafd69419b5ad88fd6a267

**Comment by nazreen** [comment]
@ryandgoulding changes made

Summary:
- renamed addPerformanceInstructions to addComputeUnitInstructions
- removed computeUnitLimitScaleFactor as a param, but kept it as a local variable inside addComputeUnitInstructions
- updated the getPrioritizationFee interface to make programId optional

**Review by nazreen — APPROVED** [review]

**Comment by nazreen** [comment]
Applied 'do not merge' as running into error when running the create task

Ran `pnpm hardhat lz:oft:solana:create --eid 30168 --program-id FzjdM63nA1DbFH7CKGUaoWvuM5kk4z3vsh44mgH6bQ7q --compute-unit-price-scale-factor 2 --only-oft-store true`

Got error

`Message: Transaction simulation failed: Attempt to debit an account but found no record of a prior credit.. `

**Comment by nazreen** [comment]
@ryandgoulding I've tested by running

`pnpm hardhat lz:oft:solana:create --eid 40168 --program-id <PROGRAM_ID> --compute-unit-price-scale-factor 2 --only-oft-store true `

and it works fine, on devnet

**Comment by nazreen** [comment]
was surprised that checks here were failing so I created a PR based off of latest `main` that only has an empty commit

it's failing too - https://github.com/LayerZero-Labs/devtools/pull/1138

**Comment by nazreen** [comment]
if after latest run of CI (after merging in main that has the fix), checks pass, Do Not Merge label should be removed.

File: examples/oft-solana/tasks/solana/index.ts
Comment (nazreen) [reply examples/oft-solana/tasks/solana/index.ts:164]: this has been fixed by https://github.com/LayerZero-Labs/devtools/pull/1091/commits/8fbdfb139dbdb73b7f2feb33636e33bf05b397ff

**Pull request merged**
